### PR TITLE
Do not use ALTS shared resource.

### PR DESCRIPTION
**Pull request by yihuazhang**
Destroy of ALTS shared nanny channel in `grpc_tsi_alts_shutdown()` in `alts_transport_security.cc` caused flakiness since at the time of this plugin destroy, some global gRPC objects (exec_ctx) that the destroy API depends on have already been destroyed beforehand. 

To resolve this flakiness, we have two options

1: do not use a shared nanny channel at all and instead, create a separate nanny channel for each ALTS TSI handshaker. For envoy use-case, we will still use the dedicated nanny channel, but it will be placed in `alts_shared_resource_dedicated` struct which will be initialized/shutdown separately by application (does not go through gRPC plugin mechanism). 

2: Add a ref-count on `alts_shared_resource` and thus on the shared nanny channel, and then ref/unref it whenever an ALTS TSI handshaker gets created/destroyed. 

After talking to Alex, I learned 1) all separate nanny channels will share a single sub-channel to the nanny which means it will not waste any nanny resources (e.g., socket fds), and have comparable performance as using a single shared nanny channel. 2) metadata server DNS resolution happens locally (i.e., does not need to talk to DNS server). So performance-wise, I did not see any obvious downside for 1) when comparing to 2).  grpclb implementation also applies the similar model as 1). 

I personally prefer strategy 1) and this PR made the corresponding changes. Please let me know if you other opinions.

---
This change is [Reviewable](https://reviewable.io/reviews/grpc/grpc/17149)

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                              FILE SIZE
 ++++++++++++++ GROWING                                                ++++++++++++++
   +14%    +115 src/core/tsi/alts/handshaker/alts_shared_resource.cc      +115   +14%
      [NEW]    +368 grpc_alts_shared_resource_dedicated_start                 +368  [NEW]
       +14%     +32 grpc_alts_shared_resource_dedicated_shutdown               +32   +14%
       +92%     +11 grpc_alts_shared_resource_dedicated_init                   +11   +92%
       +29%      +5 [Unmapped]                                                  +5   +29%
  +0.0%     +45 [None]                                                     +45  +0.0%
       +67%     +48 g_alts_resource_dedicated                                    0  [ = ]
      +0.0%     +13 [Unmapped]                                                 +45  +0.0%
  +0.1%     +16 src/core/lib/surface/call.cc                               +16  +0.1%
      [NEW]     +10 grpc_call_cancel_and_execute                               +10  [NEW]
      +1.9%      +6 [Unmapped]                                                  +6  +1.9%

 -------------- SHRINKING                                              --------------
  -3.6%    -144 src/core/tsi/alts/handshaker/alts_tsi_handshaker.cc       -144  -3.6%
     -21.5%    -208 handshaker_next                                           -208 -21.5%
     -76.2%     -16 _GLOBAL__sub_I_alts_tsi_handshaker.cc                      -16 -76.2%
  -1.2%     -64 src/core/tsi/alts/handshaker/alts_handshaker_client.cc     -64  -1.2%
     -39.7%     -52 handshaker_client_shutdown                                 -52 -39.7%
     -11.2%     -12 [Unmapped]                                                 -12 -11.2%

  -0.0%     -32 TOTAL                                                      -32  -0.0%


****************************************************************

libgrpc++.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  [ = ]       0 [None]     +48  +0.0%

  [ = ]       0 TOTAL      +48  +0.0%



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 2,019,426      Total (>)      2,019,390

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
11,180,166      Total (>)     11,180,030

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.new: 10
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.old: 10


[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                              FILE SIZE
 ++++++++++++++ GROWING                                                ++++++++++++++
   +14%    +115 src/core/tsi/alts/handshaker/alts_shared_resource.cc      +115   +14%
      [NEW]    +368 grpc_alts_shared_resource_dedicated_start                 +368  [NEW]
       +14%     +32 grpc_alts_shared_resource_dedicated_shutdown               +32   +14%
       +92%     +11 grpc_alts_shared_resource_dedicated_init                   +11   +92%
       +29%      +5 [Unmapped]                                                  +5   +29%
  +0.0%     +45 [None]                                                     +45  +0.0%
       +67%     +48 g_alts_resource_dedicated                                    0  [ = ]
      +0.0%     +13 [Unmapped]                                                 +45  +0.0%
  +0.1%     +16 src/core/lib/surface/call.cc                               +16  +0.1%
      [NEW]     +10 grpc_call_cancel_and_execute                               +10  [NEW]
      +1.9%      +6 [Unmapped]                                                  +6  +1.9%

 -------------- SHRINKING                                              --------------
  -3.6%    -144 src/core/tsi/alts/handshaker/alts_tsi_handshaker.cc       -144  -3.6%
     -21.5%    -208 handshaker_next                                           -208 -21.5%
     -76.2%     -16 _GLOBAL__sub_I_alts_tsi_handshaker.cc                      -16 -76.2%
  -1.2%     -64 src/core/tsi/alts/handshaker/alts_handshaker_client.cc     -64  -1.2%
     -39.7%     -52 handshaker_client_shutdown                                 -52 -39.7%
     -11.2%     -12 [Unmapped]                                                 -12 -11.2%

  -0.0%     -32 TOTAL                                                      -32  -0.0%


****************************************************************

libgrpc++.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  [ = ]       0 [None]     +48  +0.0%

  [ = ]       0 TOTAL      +48  +0.0%



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.new: 10
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.old: 10


[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                            FILE SIZE
 ++++++++++++++ GROWING                                              ++++++++++++++
   +14%    +115 src/core/tsi/alts/handshaker/alts_shared_resource.cc    +115   +14%
      [NEW]    +368 grpc_alts_shared_resource_dedicated_start               +368  [NEW]
       +14%     +32 grpc_alts_shared_resource_dedicated_shutdown             +32   +14%
       +92%     +11 grpc_alts_shared_resource_dedicated_init                 +11   +92%
       +29%      +5 [Unmapped]                                                +5   +29%

 -------------- SHRINKING                                            --------------
  -0.0%    -362 [None]                                               -12.6Ki  -0.1%
      -0.1%    -362 [Unmapped]                                           -12.6Ki  -0.1%
      [DEL]     -48 g_alts_resource                                            0  [ = ]
      [DEL]      -8 g_shared_resources                                         0  [ = ]
  -3.7%    -149 src/core/tsi/alts/handshaker/alts_tsi_handshaker.cc     -149  -3.7%
     -21.5%    -208 handshaker_next                                         -208 -21.5%
      [DEL]     -21 _GLOBAL__sub_I_alts_tsi_handshaker.cc                    -21  [DEL]
  [DEL]    -100 src/core/tsi/alts_transport_security.cc                 -100  [DEL]
      [DEL]     -52 grpc_tsi_alts_shutdown                                   -52  [DEL]
      [DEL]     -23 grpc_tsi_alts_init                                       -23  [DEL]
      [DEL]     -17 [Unmapped]                                               -17  [DEL]
      [DEL]      -8 alts_get_shared_resource                                  -8  [DEL]

  -0.0%    -496 TOTAL                                                -12.7Ki  -0.1%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                            FILE SIZE
 ++++++++++++++ GROWING                                              ++++++++++++++
   +14%    +115 src/core/tsi/alts/handshaker/alts_shared_resource.cc    +115   +14%
      [NEW]    +368 grpc_alts_shared_resource_dedicated_start               +368  [NEW]
       +14%     +32 grpc_alts_shared_resource_dedicated_shutdown             +32   +14%
       +92%     +11 grpc_alts_shared_resource_dedicated_init                 +11   +92%
       +29%      +5 [Unmapped]                                                +5   +29%

 -------------- SHRINKING                                            --------------
  -0.1%    -535 [None]                                               -12.7Ki  -0.1%
      -0.1%    -535 [Unmapped]                                           -12.7Ki  -0.1%
      [DEL]     -48 g_alts_resource                                            0  [ = ]
      [DEL]      -8 g_shared_resources                                         0  [ = ]
  -3.7%    -149 src/core/tsi/alts/handshaker/alts_tsi_handshaker.cc     -149  -3.7%
     -21.5%    -208 handshaker_next                                         -208 -21.5%
      [DEL]     -21 _GLOBAL__sub_I_alts_tsi_handshaker.cc                    -21  [DEL]
  [DEL]    -100 src/core/tsi/alts_transport_security.cc                 -100  [DEL]
      [DEL]     -52 grpc_tsi_alts_shutdown                                   -52  [DEL]
      [DEL]     -23 grpc_tsi_alts_init                                       -23  [DEL]
      [DEL]     -17 [Unmapped]                                               -17  [DEL]
      [DEL]      -8 alts_get_shared_resource                                  -8  [DEL]
  -5.4%     -19 src/core/plugin_registry/grpc_plugin_registry.cc         -19  -5.4%
      -5.4%     -19 grpc_register_built_in_plugins                           -19  -5.4%

  -0.0%    -688 TOTAL                                                -12.8Ki  -0.1%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 2,018,880      Total (<)      2,019,040

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
 3,971,818       Core (<)      3,972,998

11,178,553      Total (<)     11,179,739


```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.new: 10
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.old: 10


[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.new: 10
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.old: 10


[microbenchmarks] No significant performance differences
```

**Comment by markdroth** [comment]
Whoops -- nevermind that last comment; it was meant for a different PR.

**Comment by jiangtaoli2016** [comment]
Regarding to Yihua's proposals: either option is fine with me -- as long as it does not break Envoy ALTS.

**Comment by yihuazhang** [comment]
Thanks Mark and Jiangtao for the review. All comment are addressed. PTAL

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                            FILE SIZE
 ++++++++++++++ GROWING                                              ++++++++++++++
   +14%    +115 src/core/tsi/alts/handshaker/alts_shared_resource.cc    +115   +14%
      [NEW]    +368 grpc_alts_shared_resource_dedicated_start               +368  [NEW]
       +14%     +32 grpc_alts_shared_resource_dedicated_shutdown             +32   +14%
       +92%     +11 grpc_alts_shared_resource_dedicated_init                 +11   +92%
       +29%      +5 [Unmapped]                                                +5   +29%

 -------------- SHRINKING                                            --------------
  -0.1%    -503 [None]                                               -12.8Ki  -0.1%
      -0.1%    -503 [Unmapped]                                           -12.8Ki  -0.1%
      [DEL]     -48 g_alts_resource                                            0  [ = ]
      [DEL]      -8 g_shared_resources                                         0  [ = ]
  -4.5%    -181 src/core/tsi/alts/handshaker/alts_tsi_handshaker.cc     -181  -4.5%
     -20.8%    -201 handshaker_next                                         -201 -20.8%
      [DEL]     -21 _GLOBAL__sub_I_alts_tsi_handshaker.cc                    -21  [DEL]
  [DEL]    -100 src/core/tsi/alts_transport_security.cc                 -100  [DEL]
      [DEL]     -52 grpc_tsi_alts_shutdown                                   -52  [DEL]
      [DEL]     -23 grpc_tsi_alts_init                                       -23  [DEL]
      [DEL]     -17 [Unmapped]                                               -17  [DEL]
      [DEL]      -8 alts_get_shared_resource                                  -8  [DEL]
  -5.4%     -19 src/core/plugin_registry/grpc_plugin_registry.cc         -19  -5.4%
      -5.4%     -19 grpc_register_built_in_plugins                           -19  -5.4%

  -0.0%    -688 TOTAL                                                -13.0Ki  -0.1%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 2,018,868      Total (<)      2,019,028

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
 3,971,818       Core (<)      3,972,998

11,178,557      Total (<)     11,179,737


```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.new: 10
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.old: 10


[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                            FILE SIZE
 ++++++++++++++ GROWING                                              ++++++++++++++
   +14%    +115 src/core/tsi/alts/handshaker/alts_shared_resource.cc    +115   +14%
      [NEW]    +368 grpc_alts_shared_resource_dedicated_start               +368  [NEW]
       +14%     +32 grpc_alts_shared_resource_dedicated_shutdown             +32   +14%
       +92%     +11 grpc_alts_shared_resource_dedicated_init                 +11   +92%
       +29%      +5 [Unmapped]                                                +5   +29%

 -------------- SHRINKING                                            --------------
  -0.1%    -503 [None]                                               -12.8Ki  -0.1%
      -0.1%    -503 [Unmapped]                                           -12.8Ki  -0.1%
      [DEL]     -48 g_alts_resource                                            0  [ = ]
      [DEL]      -8 g_shared_resources                                         0  [ = ]
  -4.5%    -181 src/core/tsi/alts/handshaker/alts_tsi_handshaker.cc     -181  -4.5%
     -20.8%    -201 handshaker_next                                         -201 -20.8%
      [DEL]     -21 _GLOBAL__sub_I_alts_tsi_handshaker.cc                    -21  [DEL]
  [DEL]    -100 src/core/tsi/alts_transport_security.cc                 -100  [DEL]
      [DEL]     -52 grpc_tsi_alts_shutdown                                   -52  [DEL]
      [DEL]     -23 grpc_tsi_alts_init                                       -23  [DEL]
      [DEL]     -17 [Unmapped]                                               -17  [DEL]
      [DEL]      -8 alts_get_shared_resource                                  -8  [DEL]
  -5.4%     -19 src/core/plugin_registry/grpc_plugin_registry.cc         -19  -5.4%
      -5.4%     -19 grpc_register_built_in_plugins                           -19  -5.4%

  -0.0%    -688 TOTAL                                                -13.0Ki  -0.1%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 2,019,224      Total (<)      2,019,384

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
 3,972,382       Core (<)      3,973,562

11,179,113      Total (<)     11,180,297


```

File: src/core/tsi/alts/handshaker/alts_tsi_handshaker.cc [hunk 1]
```diff
@@ -418,9 +408,13 @@ tsi_result alts_tsi_handshaker_create(
...
   handshaker->options = grpc_alts_credentials_options_copy(options);
-  handshaker->base.vtable = handshaker->use_dedicated_cq
-                                ? &handshaker_vtable_dedicated
-                                : &handshaker_vtable;
+  handshaker->base.vtable =
+      use_dedicated_cq ? &handshaker_vtable_dedicated : &handshaker_vtable;
+  handshaker->channel =
+      use_dedicated_cq
+          ? nullptr
+          : grpc_insecure_channel_create(handshaker->handshaker_service_url,
```
Comment (apolcyn) [inline src/core/tsi/alts/handshaker/alts_tsi_handshaker.cc]: use of this API has <i>potentially</i> (though maybe not in practice) the same issue as the one that was fixed by https://github.com/grpc/grpc/pull/17148 - using a core surface API and initting `ExecCtx` within core?

File: src/core/tsi/alts/handshaker/alts_tsi_handshaker.cc [hunk 2]
```diff
@@ -371,6 +358,9 @@ static void handshaker_destroy(tsi_handshaker* self) {
   alts_handshaker_client_destroy(handshaker->client);
   grpc_slice_unref_internal(handshaker->target_name);
   grpc_alts_credentials_options_destroy(handshaker->options);
+  if (handshaker->channel != nullptr) {
+    grpc_channel_destroy(handshaker->channel);
```
Comment (apolcyn) [inline src/core/tsi/alts/handshaker/alts_tsi_handshaker.cc]: same comment about `ExecCtx` here as for channel creation below

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.new: 10
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.old: 10


[microbenchmarks] No significant performance differences
```

**Comment by yihuazhang** [comment]
Thanks Alex for the comments. 

Regarding `c-ares` issue, will there be any side-effect if disabling DNS queries for ALTS channel? how will the name of metadata server get resolved if not sending DNS queries? 

Regarding `exec_ctx` issue, I additionally found two other APIs (four in total) that are used in ALTS and could potentially cause the same problem: 

1) `grpc_insecure_channel_create`
2) `grpc_channel_destroy`
3) `grpc_byte_buffer_destroy`
4) `grpc_call_unref`

How about having "internal" versions of above API's as well? or is it fine to invoke them inside gRPC core. I found them being invoked in other parts of gRPC core.

**Comment by markdroth** [comment]
Yes, it would probably be a good idea to add internal versions of any C core API call that creates an `ExecCtx` that is being called from the ALTS code.  Feel free to do that in a separate PR, though.

**Comment by yihuazhang** [comment]
SG. I will work on it in a later PR then. 

Test failure: https://github.com/grpc/grpc/issues/17175

**Comment by apolcyn** [comment]
```
Regarding c-ares issue, will there be any side-effect if disabling DNS queries for ALTS channel? how will the name of metadata server get resolved if not sending DNS queries?
```

the metadata server can still be resolved from the hosts file, but c-ares will also query for load balancers with an SRV query and a service config with a TXT query

**Pull request merged**